### ISSUE TITLE
Add area intersection signals to PhysicsBody nodes

### ISSUE THREAD

**Issue by Illauriel** [godotengine/godot-proposals#8784]
### Describe the project you are working on

A third person 3D action-adventure game where physical bodies enter areas a lot.

### Describe the problem or limitation you are having in your project

TL;DR: I'd like to add the likes of area_entered/exited (and their _shape_ variants) signals to all PhysicsBodies.

Basically it all started with me implementing character swimming. To my surprise I found out there's no way to tell if `CharacterBody3D` has entered `Area3D` (representing water) from the character itself. This seems like an omission, because it's a very common use case, but currently requires clunky workarounds. It also strays from the otherwise signals-oriented design.
(Also, on `RigidBody` type nodes we have signals for collisions but not `Area` intersections.)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

`PhysicsBody` classes get 4 new signals corresponding to those on `Areas`.
`entered_area(area)`
`entered_area_shape(area_rid, area, area_shape_id, body_shape_id)`
`exited_area(area)`
`exited_area_shape(area_rid, area, area_shape_id, body_shape_id)`

This allows bodies and scripts attached to them to be aware of which areas they enter and with which shapes of those areas they intersect with minimal extra code.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

`Body` enters `Area`, they shake hands, instead of `Area` unilaterally noticing it has been entered. `Area` already does all the heavy lifting, what's left is adding calls to emit signals on the entering bodies. Signals will be emitted right after the `Area` signals. Bodies don't need to keep references to all their intersecting areas by default, and don't need to check if they are still within the area, as `Area` already does this for them, thus this enhancement should not add a significant overhead. It's should not be breaking compat.

Per @aaronfranke's suggestion the naming convention of the signals is reversed for clarity (eg. `entered_area` instead of `area_entered`), and requires adding these new names to `SceneStringNames`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Current workaround is adding a script to `Area` that would communicate with the entering body (so sort of reimplementing signals in script), but this requires adding these scripts to every area of interest in the game, which is a productivity bottleneck, and may even affect performance.
Another workaround is attaching an `Area` to the Body in question, setting its shape to the same shape the `Body` uses, and attaching this `Area`'s signals to `Body`'s script. But this is very clunky and draws an extra gizmo per body in debug view. 

### Is there a reason why this should be core and not an add-on in the asset library?

Area intersection is a part of core physics.

**Comment by KoBeWi** [comment]
When body enters area, the body is argument for the signal's callback, so you can use it to call some function:
```GDScript
func _on_body_entered(body):
	if body.is_in_group("swimmable"):
		body.enter_water(self)
```
Though until [traits](https://github.com/godotengine/godot-proposals/issues/6416) are added, it's difficult to make a "clean" solution. My example relies on dynamically typed groups. You could use static typing, but it's cumbersome with single inheritance.

**Comment by Illauriel** [comment]
> When body enters area, the body is argument for the signal's callback, so you can use it to call some function:

I actually did something like this in 3.x but as I said in the op post, it requires attaching a script to each and every Area in the game. The more complex the interactions the more of a logistics nightmare it becomes.

**Comment by KoBeWi** [comment]
Well, in the swimming example you gave I assume water is already some instance that has a script. The script does not need to be directly in Area, it can be in related node that indirectly connects the signal.

And tbh attaching scripts to functional objects is pretty normal. E.g. in my project I have a TileMap which is used in every level scene. If your node is not an instance, you can instantiate the script directly as node by dropping it on scene tree while holding Ctrl (https://github.com/godotengine/godot/pull/69015 will make it easier). Or you can make it a custom type.

**Comment by Illauriel** [comment]
> And tbh attaching scripts to functional objects is pretty normal.

I'm not arguing that it's not, but also lack of signals in the physics system feels like it's not using one of the engine's stronger points, and one of its most common workflows (click signal -> attach to method on same node). 
It might also be my personal bias, but for me Area is a thing that's "just there". It's water, wind, gravity and echo in the cave. Character is aware of environment, and environment doesn't care about character.

**Comment by ParadineDev** [comment]
I totally agree with the author of this thread. And its a common topic in our Godot chat in Telegram that it would be way more intuitive and nice to use if all bodies had collision signals. 
I switched from Unity to Godot, cause in most cases it has a way more user friendly workflow. So lack of collision signals ruined for me that feeling.

**Comment by KoBeWi** [comment]
> It might also be my personal bias, but for me Area is a thing that's "just there".

Areas are designed for detecting things that enter or exit them, hence it makes more sense that they have the signals for these events.

**Comment by Illauriel** [comment]
> Areas are designed for detecting things that enter or exit them, hence it makes more sense that they have the signals for these events.

This is why in 4x version of my prototype my workaround was to to attach another area to the CharacterBody and make its Shape the same shape the Body uses (so a capsule).
But it caused its own bunch of problems, because character capsule can change Shape when you change posture, it also changes the shape of the Area, because the Shape is shared. But the CollisionShape nodes are not shared and now I also have to write script to propagate their transform changes to the Area node. It also draws its gizmo and clutters the screen. It was a constant point of frustration for me, bad enough to go and tamper with engine code^^'

Besides, would PhysicsBodies getting signals from their very own physical world be excessive if they even get signals about being mouse-clicked?

**Comment by KoBeWi** [comment]
> my workaround was to to attach another area to the CharacterBody

How adding an area detected signal to the body would allow you to get rid of areas?

As for your workaround, I also use something similar, but that is because you can't detect 2 bodies colliding without making a physical collision (i.e. pushing).

**Comment by Illauriel** [comment]
> How adding an area detected signal to the body would allow you to get rid of areas?

I must have worded it wrong. I had an Area on a char to detect intersections with other areas. It felt annoying in many ways, esp to fit its position and orientation to that of an actual character capsule used for physics. Then I added signals and removed the Area. Life is good!
> As for your workaround, I also use something similar, but that is because you can't detect 2 bodies colliding without making a physical collision (i.e. pushing).

I'll have to struggle with pushing and pulling too at some point 😄

**Comment by KoBeWi** [comment]
> I had an Area on a char to detect intersections with other areas.

You could check body intersection with other areas and make area notify the body via the pattern I mentioned in my first comment.

idk, adding the signals in body just feels redundant to me. It creates unnecessary two-way communication.

**Comment by ParadineDev** [comment]
> > I had an Area on a char to detect intersections with other areas.
> 
> You could check body intersection with other areas and make area notify the body via the pattern I mentioned in my first comment.
> 
> idk, adding the signals in body just feels redundant to me. It creates unnecessary two-way communication.

Probably you don't realize the whole potential and pleasure of use pure signals without need to write extra scripts every time or add extra nodes/collider every time.

So if you learn some architecture patterns based on pure signals you'll discover a whole new world.

**Comment by aaronfranke** [comment]
@KoBeWi This proposal does not create two-way communication. If anything, adding the proposed signals to the body *avoids* unnecessary responsibility. The function that emits the signal on the Area node already has a reference to the body, so there is no additional responsibility. However, if the signal on the area had to be picked up by a user's GDScript so another can be emitted on the body, that adds (slight) communication complexity on the project side.

It seems like a sensible idea to allow bodies to have their logic self-contained if the user chooses to do this. The question is how common this is, because it can be worked around. One thing I am still a bit confused by is why @Illauriel iterated on their workaround by placing an Area on the character node. That seems much worse than this:

```gdscript
extends Area3D

func _on_body_entered(body: PhysicsBody3D) -> void:
	if body.has_method(&"start_swimming"):
		body.start_swimming()

func _on_body_exited(body: PhysicsBody3D) -> void:
	if body.has_method(&"stop_swimming"):
		body.stop_swimming() # This may need to be more complicated to handle adjacent areas but still that complexity would also exist with signals on the body.
```

**Comment by Illauriel** [comment]
> One thing I am still a bit confused by is why @Illauriel iterated on their workaround by placing an Area on the character node. 

I'm actually trying to teach myself out of writing custom code for everything and use built-in solutions and existing nodes as much as possible. Godot has a lot of amazing stuff out of the box and I was just ignoring it the first time around, because of my habits carried over from work in other engines. So now I'm trying to do things "the Godot way". Attaching an Area to the character to detect other areas is actually a thing I saw suggested online, and decided to try^^' It's an actual workaround people use to stay within the "signals for everything" paradigm.

**Comment by KoBeWi** [comment]
> This proposal does not create two-way communication.

By two-way communication I meant that when a body enters area, both body and the area will basically emit the same signal. It's redundant.

> It seems like a sensible idea to allow bodies to have their logic self-contained if the user chooses to do this.

idk, this is probably a matter of taste, but to me such logic feels backwards.
Water detects body -> body is player -> notify player. Spikes detect body -> body is player -> deal damage to player.
vs
Area enters player -> check if area is water -> do water logic. Area enters player -> check if area is spikes -> deal damage to self.

Also, never in my code I had an area where reversing the signals would allow me to remove its script, because such areas usually have more logic attached. The closes one would be my PlayerDetector object, which only checks if the body entered is a player and passes the signal further, but even there I added extra logic to optionally allow the detector work only once and auto-free.

**Comment by ParadineDev** [comment]
> Also, never in my code I had an area where reversing the signals would allow me to remove its script, because such areas usually have more logic attached. The closes one would be my PlayerDetector object, which only checks if the body entered is a player and passes the signal further, but even there I added extra logic to optionally allow the detector work only once and auto-free.

Maybe you don't need this, but some people definitely do. That's why I don't understand why some people try to force anyone from having good time with the engine. It will not break your experience, but it will significantly improve mine.

**Comment by aaronfranke** [comment]
> I'm actually trying to teach myself out of writing custom code for everything and use built-in solutions and existing nodes as much as possible.

The area node's body entered system is a built-in solution. You writing a few lines to forward it is not a big deal. It's a simpler, cleaner, easier, and more efficient approach than adding extra nodes to your tree.

> idk, this is probably a matter of taste, but to me such logic feels backwards.

When possible, it's best for objects to manage themselves. Sure, in some cases areas should manage their own effect and alter bodies. But in other cases it's better for bodies to manage what they are affected by.

Godot's own internal code has places where it prefers bodies to keep track of areas, not areas keeping track of bodies. For example, with gravity, the body needs to look through all area nodes it's overlapping and apply the gravity in the order determined by the area priorities. Trying to do this with areas applying stuff to bodies simply would not work, because the areas are not aware of all other areas influencing the body. This proposal is asking to make it easier to follow a similar pattern in user scripts if the users need this pattern.

> Also, never in my code I had an area where reversing the signals would allow me to remove its script, because such areas usually have more logic attached.

I agree here, if there is a "water area", it's likely that will need more logic anyway, so the workaround I mentioned in my post is just a few lines added to an existing script.

**Comment by Illauriel** [comment]
> Water detects body -> body is player -> notify player

The proposed PR does basically this, except it does so in CPP, doesn't do it's own validation of who's allowed to be notified, and allows to create a method in receiving body's script in three clicks. Also in "Water detects body" case there's this duality that @aaronfranke mentions above - Areas can represent two separate concepts at the same time - an active sensor, like a security camera or a tripwire, and environmental conditions. For the latter, I think it's better if the area doesn't know anything about the player and its methods and is agnostic to whatever enters it.

> Area enters player -> check if area is water

That's why the signal names are reversed in the PR (entered_area instead of area_entered) so it's Player enters area -> check if area is water. It allows player physics to be self-contained, and keeps all the relevant logic within the player scene file.

> Godot's own internal code has places 

And in the end it all amounts to looking through the crystal ball into the world of PhysicsServer's internal representaion. 😄 

> if there is a "water area", it's likely that will need more logic anyway

Tbh in my case I only need Area's water level, density and viscosity, and I calculate buoyancy with a simplified formula in CharacterBody. And since I'm using a C++ hydro module for other water physics, these properties are built into WaterArea custom node, and I don't have scripts attached to it^^'

**Comment by aaronfranke** [comment]
> these properties are built into WaterArea custom node, and I don't have scripts attached to it^^'

The same principle for the workaround applies in the C++ code for your WaterArea:

```cpp
WaterArea::WaterArea() {
	connect("body_entered", callable_mp(this, &WaterArea::_on_body_entered));
	connect("body_exited", callable_mp(this, &WaterArea::_on_body_exited));
}

void WaterArea::_on_body_entered(Node *body) {
	if (body->has_method("start_swimming")) {
		body->call("start_swimming");
	}
}

void WaterArea::_on_body_exited(Node *body) {
	if (body->has_method("stop_swimming")) {
		body->call("stop_swimming");
	}
}
```

**Comment by dalexeev** [comment]
I'm also skeptical about this, at least because of the implementation in linked PR: body signals are not emitted by the body itself, but by the area the body enters/exits. In my opinion, emitting a signal from outside the object is an antipattern. Signals should report events that happened to the object itself (inside it), and not about external influences. (It is acceptable in user code to emit signals from another object if it is a "component" of the object to which the signal belongs. But this also does not count as emission from outside.) The solution you propose contradicts the Observer pattern and the way signals are used in Godot.

Also, in my opinion, it is bad that emitting (or not emitting) of a body signal depends on the settings of the area (`monitoring`), not the body.

Note that you can make a body behave like an area if you use `PhysicsDirectSpaceState{2D,3D}` and/or a physics server. This is less efficient than an area if you call methods in each frame, but is more efficient if you only need to check for intersections occasionally, such as during interactions. I believe that in this case it will work even if area's `monitoring` is false, unlike your solution.

Moreover, this distinction between bodies and areas was originally built into physics, at the level of the physical server, not the nodes. Bodies can only detect collisions with other bodies, areas can detect entry, exit, and intersections with bodies and other areas. The addition of these artificial signals attempts to hide this fundamental difference without eliminating the fundamental differences between bodies and areas. (Disclaimer: I'm not a physics contributor, so I may be wrong here.)

Finally, this affects performance as it doubles the number of signals emitted.

I understand that there is often some confusion between what to use (bodies or areas) and how best to organize the physics to avoid spaghetti code. But in my opinion, this proposal only increases this confusion, artificially blurring the borders between bodies and areas. It's entirely possible we're dealing with [XY problem](https://en.wikipedia.org/wiki/XY_problem), and the real problem isn't where you're trying to solve it. What we need is more detailed documentation, tutorials, and best practices, rather than these kinds of "workarounds" that allow everyone to code the way they want, at the expense of the intended architecture and performance.

**Comment by Illauriel** [comment]
> In my opinion, emitting a signal from outside the object is an antipattern.

 This is a valid concern, although, since it's happening at the source level, it will be obscured to the end user.
 
> Signals should report events that happened to the object itself (inside it), and not about external influences

![Screenshot 2024-01-04 152936](https://github.com/godotengine/godot-proposals/assets/10573379/9d392cb3-43f9-45b8-8d9b-e5d78f4739bb)
This image will be haunting this discussion with its ominous presence. In this case a viewport is calling a method on collision_object_3d, that in turn emits a signal. Would changing my PR so that Area calls a method on PhysicsBody alleviate your concern?

> The solution you propose contradicts the Observer pattern and the way signals are used in Godot.

The most basic and default way signals are used in Godot is "Open Node Signals tab -> click signal -> click Connect". Currently there's no way to have this most basic workflow on the Body. Would connecting area signals from the Area to the Body in question (eg character) be a cleaner, more platonic solution? :)

> Also, in my opinion, it is bad that emitting (or not emitting) of a body signal depends on the settings of the area (monitoring), not the body.

A non-monitoring Area is effectively disabled. I would argue that detecting entry into non-monitoring areas might be undesirable, but now I'm curious how areas behave when a monitoring one and a non-monitoring one intersect. Will check it later^^

> if you use PhysicsDirectSpaceState{2D,3D} and/or a physics server

Sure. And, maybe it would be better to inform objects of their intersections/collisions at the server level, but I don't know its internal workings that well yet. Also, no average user will go that deep for pretty basic use-cases like detecting intersections.

> Moreover, this distinction between bodies and areas was originally built into physics, at the level of the physical server

If this is an intentional an explicit initial design choice, I would concede, although it really doesn't make things easier or more clear from UX standpoint.

> the real problem isn't where you're trying to solve it.

My problem is not having signals. I want signals. The specific implementation is irrelevant as long as I get signals.

**Comment by AThousandShips** [comment]
I'd agree with the already presented points that this isn't really how `Area`s and `Body`s are designed or usually used

In my experience, my own code, and in how it's done in examples, `Areas` handle behaviour based on interaction, opens a door, shows a message, teleports you, in most of these cases, near all I'd say, the logic belongs on the `Area`. Not only is it contained there, i.e. specific, like what door to open, where to teleport you, etc., but the `Body` doesn't need to handle it all, it doesn't make sense to have the logic on the player node, in cases where you do want the `Body` to do the logic you just delegate to it, but that's the *exception* not the rule

This hurts performance, assumes a lot about the bodies that the current code *does not*, and has various pitfalls and lots of complicated logic with how and when to interact

I think this goes against how `Area` is designed and intended, which is why this feature wasn't added long ago, unlike the other signals

**Comment by Illauriel** [comment]
TIL Areas can be Monitoring and Monitorable, which roughly corresponds to my assumption that Areas can represent two separate concepts: 
active triggers
> Areas handle behaviour based on interaction, opens a door, shows a message, teleports you

and environmental modifiers
 
> a thing that's "just there". It's water, wind, gravity and echo in the cave. Character is aware of environment, and environment doesn't care about character.

My currently proposed solution only works if Area is Monitoring, as does the 'just put a script on the Area' workaround. If there are signals for intersecting with Areas on a Body, I'd expect the opposite behavior - detecting intersections if Monitorable is on. This also highlights the advantage of the 'put an Area on the body to detect other Areas' workaround, that I haven't thought about previously. It's clunky but would detect Monitorable areas that are themselves not Monitoring (so are not a sensor for activating behaviors).

> I'd agree with the already presented points that this isn't really how `Area`s and `Body`s are designed or usually used

I still argue that nothing beats setting up behaviors in 5 seconds and 3 clicks.
https://github.com/godotengine/godot-proposals/assets/10573379/8edd0806-29f1-42b3-88f0-d35671f538fb

> This hurts performance

No more than attaching a script to do basically the same thing to every Area 🤷‍♂️ (Also the script, as we've just found out, doesn't produce the intended behavior anyway.)

**Comment by AThousandShips** [comment]
> No more than attaching a script to do basically the same thing to every Area 🤷‍♂️ (Also the script, as we've just found out, doesn't produce the intended behavior anyway.)

No, emitting an additional signal isn't free, nor are the checks for them

> I still argue that nothing beats setting up behaviors in 5 seconds and 3 clicks.

I mean you still write code, and still have a signal, there's no less work with this

**Comment by dalexeev** [comment]
The problem is that you're adding responsibility to the class for something it's not really responsible for. Areas can detect the entrance/exit of bodies/areas, but bodies cannot. You are making a "little lie" by making the areas responsible for the bodies. But quite often you have to pay for it, sooner or later. Therefore, I suggest to wait for comments from physics contributors who are familiar with the architecture.

Programming in general and OOP in particular is modeling. When designing models, they have certain traits that determine the limits of applicability. Not all problems can be solved equally well using the same model. Adding even small previously unaccounted inputs sometimes requires a significant revision of the model. As well unintended use of the model leads to problems when working with it.

We need qualified comments both on the proposal (how it corresponds to the architecture and bodies vs areas differences), and on the proposed solution, if the proposal is correct. Please note that the physical system consists of several levels (servers and nodes), and it is necessary to determine the correct level on which this should be implemented (if the proposal is correct). It might be better to implement this at the lower level rather than adding a patch at the top level.

> My problem is not having signals. I want signals. The specific implementation is irrelevant as long as I get signals.

This is not an constructive position. I mentioned XY problem: body signals cannot be a primary problem, you need them to solve some practical problems. Which are worth discussing first to understand whether it is possible to solve them without adding new inputs with unknown consequences. You mentioned that signal connections in the editor are more convenient for you, but why should a body be responsible for detection rather than an area?

**Comment by Illauriel** [comment]
> the new signals (plural) are emitted no matter what and has a cost even if nothing is connected to them

This is a valid concern, I agree.

> I believe the use case is perfectly justified, but the proposed solution has pitfalls and the work-around is quite simple.

Now that the monitoring/monitorable issue has turned up, it totally invalidates my proposed solution. I'll close the PR and edit the opening post.

> Therefore, I suggest to wait for comments from physics contributors who are familiar with the architecture.

I agree, and as I said above, my solution is not a valid one, not only because of the responsibility issues you've mentioned, but also because it will not work with non-monitoring, yet monitorable areas. This problem also persists in the simplest workaround - attaching a script that would have Area filter incoming bodies and calling methods on relevant bodies, that @KoBeWi suggested initially.

> It might be better to implement this at the lower level rather than adding a patch at the top level.

Most likely so.

> body signals cannot be a primary problem, you need them to solve some practical problems. Which are worth discussing first to understand whether it is possible to solve them without adding new inputs with unknown consequences. 

Yeah, now that the first implementation is out of the way, we can discuss the conceptual issues and come to some kind of conclusion of how things should work.

> You mentioned that signal connections in the editor are more convenient for you, but why should a body be responsible for detection rather than an area?

With further thought given to this topic, I'd say main point for why Body should be able to detect Area intersections, is that Areas can be Monitorable while not Monitoring. I think that bodies should still be able to interact with such areas, and they already partially can eg. a non-monitoring Area still contributes to PhysicsDirectBodyState.total_gravity. But it doesn't for other properties like wind, and whatever custom properties you might want to assign to it (wink-wink, buoyancy stuff) Exposing more area properties via PhysicsServer is a solution suitable for advanced users, but imo has some accessibility problems.

To summarize my takes from the previous posts and some new ones:
- Signal-based workflows are very popular with the community
- Users coming from certain other engines expect both interacting collision objects to receive intersection callbacks
- "Have Area set things on body" only works if Area is Monitoring. This is not desirable
- "Attach an Area to body to detect other Areas" is, surprisingly, a better workaround, because it respects non-monitoring Areas
- Our initial discussions didn't take the conceptual distinction between Monitorable/Monitoring (smart/dumb, sensor/environment) areas. @KoBeWi and @AThousandShips expect Areas to work as sensors, I expected areas to work as environment, but apparently they can represent both, and very robustly at that.

**Comment by KoBeWi** [comment]
> I'd say main point for why Body should be able to detect Area intersections, is that Areas can be Monitorable while not Monitoring. I think that bodies should still be able to interact with such areas

Monitorable non-monitoring areas can still be detected by other areas (or at least should be, IIRC something was bugged). But yeah, bodies interacting with monitorable-only areas is a use-case not supported currently (whether it's desirable/useful is another thing).

**Comment by AThousandShips** [comment]
It's not so much that it's wrong to use areas like this, it's that it's unconventional in my experience and how they are presented in official material, so adding code to handle that is hard to defend IMO, because we normally support the usual behaviour that most people use, as they don't make sense to build yourself, whereas unconventional things are open to be implemented as you see fit

**Comment by Illauriel** [comment]
>  so adding code to handle that is hard to defend IMO, because we normally support the usual behaviour that most people use

I decided to have a look which behaviour people actually use, because tbh I've never watched area-related tutorial videos, and I've seen both "add script to area, to nudge body when it enters" and "put area on player, maybe even 2 or 3 for the good measure". But my good Lord, there are also people that are connecting Area's enter signals to the player's body... So not even "calling methods on enter". They just hotwire the signal to the player. That one signal that would return player's own body as an argument. How's that for an anti-pattern?

I actually agree that it's best to use existing solutions, and that workarounds are okay, and preferable as long as they don't pose a bunch of problems themselves. And of course, introducing stuff to core requires to think of performance. I really didn't expect my initial solution to turn out so costly in this regard.

I think docs should really be updated. Eg. Area and Signal tutorials never mention how to _properly_ handle area-body interactions. But first it would be good to agree on what "properly" means. Maybe having a bunch of areas on a character is a good thing. Maybe non-monitoring areas are not meant to be interacted with and only contribute physics overrides (btw wind settings need some love).

**Comment by ParadineDev** [comment]
I have found the perfect solution that will suit everyone at once.

Some people, including me, want that all Bodies should have signals that check for collisions.

Other people are against it, because collisions are checked in Area.

So let's just let any Body check collisions by itself. Then the signals will be fine - they will be appropriate.

And to avoid unnecessary checks, we have layers, which should be used.
